### DISCUSSION TITLE
단축 url 생성 패턴에 대해서 질문드립니다.

### DISCUSSION THREAD

**Discussion by trueS2** [100-hours-a-week/discussions#68] [개발] [answered]
현재 프로젝트에서 게시글이 생성되면 동시에 단축 URL 주소가 랜덤하게 생성되도록 구현하고 있습니다. 그런데 현재 방식은 URL이 너무 길고 복잡하게 만들어지는 것 같아서, 단축 URL의 가독성을 개선하고자 합니다.

단축 URL을 생성할 때, 예를 들어 "길이 15자 이하", "닉네임 + 숫자" 등과 같이 가독성 좋은 패턴을 적용하려고 하는데요. 단축 URL 생성 시 좋은 패턴이나 일반적으로 많이 사용하는 방식이 있을까요?

**Comment by ej31** [comment, marked as answer]
# 답변

- 단축 URL의 목적을 생각하면 저라면 크게 세 가지로 적용할 것 같아요
  1. 해쉬값 방식
  2. 무작위로 선정 된 단어 조합 방식
  3. 유저 커스텀 방식
- 하나씩 톺아봅시다. 🤡

## 해쉬값 방식
- 많은 단축 URL이 이 방식을 채택한다고 생각합니다.
- 다만, 자릿수가 너무 짧은 해쉬 값은 충돌의 위험이 있기 때문에 (=중복의 위험이 있기 때문에) 비즈니스 로직 안에서 중복되는 URL 인지 꼭 검사를해야 하고 2차 안전장치로 DB 에 유니크도 걸어놔야 합니다.
- 반대로 긴 해쉬값은 유저의 가독성이 좋지 않지만 겹칠 위험이 상대적으로 적겠지요. (유저의 가독성을 희생해서 구현 난이도를 낮추는 트레이드 오프)

## 무작위로 선정 된 단어 조합 방식
- 이게 무엇이냐면.. 우선 이 [링크](https://jimpix.co.uk/words/username-experiments-page-combinations.php#title)의 페이지에 들어가서 새로고침을 여러번해보세요.
- 그러면 단어가 막 바뀌는게 보이실겁니다. 이 단어를 하이픈(`-`)으로 엮고 소문자로 누르면 완성입니다.
- 이러한 단어 조합을 해주는 함수를 만들어서 해당 조합을 단축 URL 값으로 사용하는 것이지요
- 예를 들어 단축 URL이 `https://connectrip.kro.kr/short/label-partner-downlight` 와 같이 생성 되는 것이지요
> 사실 단축 URL 구현할거면 별도로 단축 URL 용 도메인 구매해서 따로 관리해야 하지만 여기선 논외로 하겠습니다 🤡
- 해시로 구현하는 것보단 상대적으로 비즈니스 로직이 조금 더 복잡하겠지만 유저의 가독성은 어느정도 확보 될 거라고 생각됩니다.
  - 랜덤한 단어를 어디서 가져오냐면.. `word dictionary github` 로 구글에 검색해보시면 다양한 리포지토리가 나옵니다.
  - 이 리토지토리에 나와 있는 단어를 전처리 하셔야 합니다. (전처리는 쓰면 안되는 단어, 불필요한 부분 등은 모두 지우고 필요한 건 추가하고.. 그런거 하는 겁니다.) 
  - 그리고 이 단어를 파일에 넣고 그대로 쓰면 안되고 DB 에 캐싱해둔채로 써야 합니다. 안그러면 읽을 때마다 File I/O 가 발생하기 때문이지요.

- 난이도가 올라가긴 하지만 유저의 가독성을 얻을 수 있습니다. (유저의 가독성을 확보될 수 있지만 구현 난이도가 올라가는 트레이드 오프)

## 유저 커스텀 방식
- 가장 불지옥입니다.
- 유저는 언제나 최악의 악동이자 가장 필요한 존재이기 때문입니다.
- 이것 역시 그냥 만들면 되는거 아니냐고 생각하시겠지만 욕설, 성적인단어, 비하단어 등 필터를 만들어야 합니다.
- 또한 특수문자 필터링과 base62 encode 등을 고려했을 때 난이도가 훅 올라가게 됩니다.
- 대신 유저의 가독성은 최대치를 찍겠지요. (유저 가독성은 최대치이지만 구현 난이도 또한 최대치를 찍습니다. 역시나 둘 사이의 트레이드 오프)

# 결론
- 엔지니어의 모든 결정은 언제나 _Trade-Off_ 입니다.
- 시간이 부족한 경우엔 우선 구현할 수 있는 방법을 선택하는 것이 답이 될 수 있습니다.
- 잘 생각해보시고.. 하고 싶은것과 해야 하는것을 구분해서 선택하시면 될 것 같습니다.
- 굿 럭 👴
